scale-decode/examples/enum_decode_as_type.rs

```diff
@@ -29,32 +32,39 @@ enum Foo {
 }
 
 // Define a struct that will be our `Visitor` capable of decoding to a `Foo`.
-struct FooVisitor;
+struct FooVisitor<R>(PhantomData<R>);
+
+impl<R> FooVisitor<R> {
+    fn new() -> Self {
+        Self(PhantomData)
+    }
+}
 
 // Describe how to turn `Foo` into this visitor struct.
 impl IntoVisitor for Foo {
-    type Visitor = FooVisitor;
-    fn into_visitor() -> Self::Visitor {
-        FooVisitor
+    type AnyVisitor<R: TypeResolver> = FooVisitor<R>;
+    fn into_visitor<R: TypeResolver>() -> Self::AnyVisitor<R> {
```

Comment:
DQ: This `TypeResolver` thingy is it implemented for all types in `PortableTypeRegistry` or how does it work?

For instance if I have some custom type is it possible that I need to implement `TypeResolver` for it? Is it simple to implement?

Reply:
I guess it's unlikely that one needs to implement the `TypeResolver` trait but would good know when the `T: !TypeResolver` :)

Reply:
Yup, so the idea here is that any type which wants to implement `IntoVisitor` must do so in a way that is generic over which `TypeResolver` to use.

And then, for V14 metadata, `scale_info::PortableRegistry` is a valid choice of type resolver (it has all of the info needed to encode/decode types). For metadata before V14, users need to provide their own `TypeResolver` because we're not given any type info. For that resolver we'll have a map from the type names found in old metadata to the relevant info needed to encode/decode the type. The `scale-info-legacy` crate will define this type resolver and probably also have a load of default type mappings so that users don't need to define _everything_ themselves (just type names unique to specific chains).